project/tickets/views.py

```diff
@@ -0,0 +1,312 @@
+    voters = [vote["user_id"] for vote in upvotes]
+    donors = User.objects.filter(id__in=voters)
+    # render drop-down list for superusers to update status
+    ticket_status_list = TicketStatus.objects.all()
+    tkt_status = ticket.ticket_status.id
+    # POST methods
+    if request.method == "POST":
+        comment_form = CommentForm(request.POST)
+        if comment_form.is_valid():
+            comment_form.instance.commenter = request.user
+            comment_form.instance.ticket = ticket
+            comment_form.save()
+            # remove two ticket views to avoid duplicates on POST
+            ticket.views -= 2
+            ticket.save()
+            messages.success(
+                request, f"Your comment was added!")
+            return redirect(tickets_view_one, ticket.pk)
+    else:
+        comment_form = CommentForm()
+        donation_form = DonationForm()
+    context = {
+        "comment_form": comment_form,
+        "comments": comments,
+        "donation_form": donation_form,
+        "donors": donors,
+        "publishable": settings.STRIPE_PUBLISHABLE,
+        "ticket": ticket,
+        "ticket_status_list": ticket_status_list,
+        "tkt_status": tkt_status,
+        "voters": voters,
+    }
+    return render(request, "tickets_view_one.html", context)
+
+
+@login_required
+def tickets_edit(request, pk):
+    """ Edit a Single Ticket """
+    ticket = get_object_or_404(Ticket, pk=pk)
+    if request.method == "POST":
+        ticket_form = TicketForm(request.POST, instance=ticket)
+        if ticket_form.is_valid():
+            ticket_form.instance.date_edited = timezone.now()
+            ticket_form.save()
+            messages.success(
+                request, f"Ticket successfully updated!")
+            return redirect(tickets_view_one, ticket.pk)
+    else:
+        ticket_form = TicketForm(instance=ticket)
+    context = {
+        "ticket": ticket,
+        "ticket_form": ticket_form,
+    }
+    return render(request, "tickets_edit.html", context)
+
+
+@login_required
+def tickets_delete(request, pk):
+    """ Delete a Ticket """
+    ticket = get_object_or_404(Ticket, pk=pk)
+    ticket.delete()
+    messages.success(
+        request, f"Your ticket has been deleted!")
+    return redirect(tickets_view_all)
+
+
+@login_required
+def upvote_add(request, pk):
+    """ Upvote a Ticket """
+    ticket = get_object_or_404(Ticket, pk=pk)
+    ticket.upvotes += 1
+    # remove a ticket view to avoid duplicates
+    ticket.views -= 1
+    ticket.save()
+    if request.method == "POST":
+        donation_form = DonationForm(request.POST)
+        if donation_form.is_valid():
+            # amount to pay / donate
+            donation_amount = 0
+            donation_amount += int(request.POST.get("donation_amount"))
+            try:
+                # build Stripe payment
+                token = request.POST["stripeToken"]
+                customer = stripe.Charge.create(
```

Comment:
DRY. Put into a function and reuse.

Reply:
I can't seem to get this to work. It isn't 100% identical from the 'new feature' vs 'upvote feature', so I can only attempt the *try/except* into its own function, however it keeps failing, saying the variable for Stripe (*customer*) is not recognizable. Since the try/except is the only part that's identical, I'm just going to keep it as-is for the time being, and can re-visit after submission when I have more time.